actfw_core/task/producer.py

```diff
@@ -1,17 +1,71 @@
-from .pipe import Pipe
+import traceback
+from queue import Empty, Full
+from typing import Generic, List, TypeVar
 
+from ..util.pad import _PadBase, _PadBlocking, _PadIn
+from .consumer import _ConsumerMixin
+from .task import Task, _TaskI
 
-class Producer(Pipe):
+T_OUT = TypeVar("T_OUT")
 
-    """Producer Task."""
 
-    def __init__(self):
+class _ProducerMixin(Generic[T_OUT], _TaskI):
+    out_queues: List[_PadIn[T_OUT]]
+    out_queue_id: int
+
+    def __init__(self) -> None:
         """"""
-        super().__init__()
+        self.out_queues = []
+        self.out_queue_id = 0
+
+    def _add_out_queue(self, q: _PadIn[T_OUT]) -> None:
+        self.out_queues.append(q)
 
-    def _add_in_queue(self, q):
-        raise NotImplementedError("This is producer")
+    def _new_pad(self) -> _PadBase[T_OUT]:
+        return _PadBlocking()
 
-    def _inlet(self):
+    def connect(self, follow: _ConsumerMixin[T_OUT]) -> None:
+        """
+        Connect following task.
+        """
+        assert isinstance(follow, _ConsumerMixin)
```

Comment:
Maybe a redundant assertion after mypy is introduced?

Reply:
IIUC, no.
Type-hint does not check types in runtime and we cannot assume that users of the library use mypy.
It is convenient for users to show the error message "AssersionError" rather than "object foo does not have method `_add_in_queue()`. I think we do not have to assert everywhere, but we should if the code seems to lead to mistakes of lots of users.

Reply:
right